https://drive.google.com/file/d/1etGP85wJWsgqbsh3gX5oxc8dPKLMA1eT/view

https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv

https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt

https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv

https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
raw_data1 = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep="\t")
raw_data2 = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt", sep=" ")
raw_data3 = pd.read_csv("https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv")
raw_data4 = pd.read_csv("https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv")

# Analyzing raw_data1

In [3]:
data1 = raw_data1.copy()

In [4]:
data1.head(20)

order_id  quantity                              item_name  \
0          1         1           Chips and Fresh Tomato Salsa   
1          1         1                                   Izze   
2          1         1                       Nantucket Nectar   
3          1         1  Chips and Tomatillo-Green Chili Salsa   
4          2         2                           Chicken Bowl   
5          3         1                           Chicken Bowl   
6          3         1                          Side of Chips   
7          4         1                          Steak Burrito   
8          4         1                       Steak Soft Tacos   
9          5         1                          Steak Burrito   
10         5         1                    Chips and Guacamole   
11         6         1                   Chicken Crispy Tacos   
12         6         1                     Chicken Soft Tacos   
13         7         1                           Chicken Bowl   
14         7         1                    Chips and Guacamole   
15         8         1  Chips and Tomatillo-Green Chili Salsa   
16         8         1                        Chicken Burrito   
17         9         1                        Chicken Burrito   
18         9         2                            Canned Soda   
19        10         1                           Chicken Bowl   

                                   choice_description item_price  
0                                                 NaN     $2.39   
1                                        [Clementine]     $3.39   
2                                             [Apple]     $3.39   
3                                                 NaN     $2.39   
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                 NaN     $1.69   
7   [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8   [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9   [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25   
10                                                NaN     $4.45   
11  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
12  [Roasted Chili Corn Salsa, [Rice, Black Beans,...     $8.75   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
14                                                NaN     $4.45   
15                                                NaN     $2.39   
16  [Tomatillo-Green Chili Salsa (Medium), [Pinto ...     $8.49   
17  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...     $8.49   
18                                           [Sprite]     $2.18   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75

In [5]:
data1["choice_description"].values

array([nan, '[Clementine]', '[Apple]', ...,
       '[Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]',
       '[Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]',
       '[Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]'],
      dtype=object)

## Handling the item_price

In [6]:
# remove $
data1["item_price"] = data1["item_price"].str.replace('$', '').astype("float64")

# Handeling choice description

In [7]:
# convert to original list
data1["choice_description"] = data1["choice_description"].fillna("[]")
lst_series = (data1["choice_description"].str.strip('][').str.split(', '))
# data1["choice_description"] = lst_series
# Finding unique list of items
unique_items = list(set(lst_series.sum()))

# Just modifying the data for visual purpose
final_unique_items = []
for element in unique_items:
    if element.startswith('['):
        element = element.replace('[', '')
    if element.endswith(']'):
        element = element[::-1].replace(']', '')[::-1]
    
    if element != "":
        final_unique_items.append(element)

print("Total no. of dishes:", len(final_unique_items))

# Now, we got got our final_unique_items which each one of the customer ordered, let's create a new column for them
data1[final_unique_items] = 0


# If the customer ordered one of them, we replace the column value by 1, otherwise 0


present_items = [[True if item in item_list else False for item_list in data1["choice_description"].values ] for item in final_unique_items]

for item, validity in zip(final_unique_items, present_items):
    data1.loc[validity, item] = 1



Total no. of dishes: 60


# Replacing choice_description column with the net items under it

In [8]:
data1.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description', 'item_price',
       'Brown Rice', 'Black Beans', 'Fajita Vegetables', 'Braised Barbacoa',
       'Tomatillo-Red Chili Salsa (Hot)', 'Fresh Tomato Salsa (Mild)', 'Salsa',
       'Guacamole', 'Fresh Tomato Salsa', 'Fajita Veggies', 'Pinto Beans',
       'Sour Cream', 'Apple', 'Rice', 'Coca Cola',
       'Tomatillo Green Chili (Medium)', 'Lettuce', 'White Rice', 'Lemonade',
       'Roasted Chili Corn (Medium)', 'Pineapple Orange Banana',
       'Roasted Chili Corn Salsa', 'Roasted Chili Corn Salsa (Medium)',
       'Nestea', 'Diet Dr. Pepper', 'Adobo-Marinated and Grilled Steak',
       'Grapefruit', 'Blackberry', 'Peach Orange', 'Diet Coke', 'Dr. Pepper',
       'Cilantro-Lime Rice', 'Fresh Tomato (Mild)',
       'Adobo-Marinated and Grilled Chicken', 'Tomatillo Red Chili (Hot)',
       'Tomatillo Red Chili Salsa', 'Tomatillo Green Chili Salsa',
       'Pomegranate Cherry', 'Clementine',
       'Tomatillo-Green Chili Sals

In [9]:
# Now, let's drop the choice_description column

choice_description_count = data1[final_unique_items].sum(axis=1)

data1.insert(loc = 4,
          column = 'choice_description_count',
          value = choice_description_count)

In [10]:
data1.head(20)

order_id  quantity                              item_name  \
0          1         1           Chips and Fresh Tomato Salsa   
1          1         1                                   Izze   
2          1         1                       Nantucket Nectar   
3          1         1  Chips and Tomatillo-Green Chili Salsa   
4          2         2                           Chicken Bowl   
5          3         1                           Chicken Bowl   
6          3         1                          Side of Chips   
7          4         1                          Steak Burrito   
8          4         1                       Steak Soft Tacos   
9          5         1                          Steak Burrito   
10         5         1                    Chips and Guacamole   
11         6         1                   Chicken Crispy Tacos   
12         6         1                     Chicken Soft Tacos   
13         7         1                           Chicken Bowl   
14         7         1                    Chips and Guacamole   
15         8         1  Chips and Tomatillo-Green Chili Salsa   
16         8         1                        Chicken Burrito   
17         9         1                        Chicken Burrito   
18         9         2                            Canned Soda   
19        10         1                           Chicken Bowl   

                                   choice_description  \
0                                                  []   
1                                        [Clementine]   
2                                             [Apple]   
3                                                  []   
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...   
6                                                  []   
7   [Tomatillo Red Chili Salsa, [Fajita Vegetables...   
8   [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...   
9   [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...   
10                                                 []   
11  [Roasted Chili Corn Salsa, [Fajita Vegetables,...   
12  [Roasted Chili Corn Salsa, [Rice, Black Beans,...   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...   
14                                                 []   
15                                                 []   
16  [Tomatillo-Green Chili Salsa (Medium), [Pinto ...   
17  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...   
18                                           [Sprite]   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...   

    choice_description_count  item_price  Brown Rice  Black Beans  \
0                          0        2.39           0            0   
1                          1        3.39           0            0   
2                          1        3.39           0            0   
3                          0        2.39           0            0   
4                         12       16.98           0            1   
5                         15       10.98           0            0   
6                          0        1.69           0            0   
7                         17       11.75           0            1   
8                         11        9.25           0            0   
9                         15        9.25           0            1   
10                         0        4.45           0            0   
11                        13        8.75           0            1   
12                        11        8.75           0            1   
13                        13       11.25           0            0   
14                         0        4.45           0            0   
15                         0        2.39           0            0   
16                        10        8.49           0            0   
17                        15        8.49           0            1   
18                         1        2.18           0            0   
19                        13        8.75           0        

In [11]:
data1.describe()

order_id     quantity  choice_description_count   item_price  \
count  4622.000000  4622.000000               4622.000000  4622.000000   
mean    927.254868     1.075725                  7.633059     7.464336   
std     528.890796     0.410186                  6.279394     4.245557   
min       1.000000     1.000000                  0.000000     1.090000   
25%     477.250000     1.000000                  0.000000     3.390000   
50%     926.000000     1.000000                  9.000000     8.750000   
75%    1393.000000     1.000000                 13.000000     9.250000   
max    1834.000000    15.000000                 24.000000    44.250000   

        Brown Rice  Black Beans  Fajita Vegetables  Braised Barbacoa  \
count  4622.000000  4622.000000        4622.000000       4622.000000   
mean      0.000433     0.291000           0.156209          0.000433   
std       0.020800     0.454272           0.363093          0.020800   
min       0.000000     0.000000           0.000000          0.000000   
25%       0.000000     0.000000           0.000000          0.000000   
50%       0.000000     0.000000           0.000000          0.000000   
75%       0.000000     1.000000           0.000000          0.000000   
max       1.000000     1.000000           1.000000          1.000000   

       Tomatillo-Red Chili Salsa (Hot)  Fresh Tomato Salsa (Mild)  \
count                      4622.000000                4622.000000   
mean                          0.056036                   0.075941   
std                           0.230017                   0.264933   
min                           0.000000                   0.000000   
25%                           0.000000                   0.000000   
50%                           0.000000                   0.000000   
75%                           0.000000                   0.000000   
max                           1.000000                   1.000000   

             Salsa    Guacamole  Fresh Tomato Salsa  Fajita Veggies  \
count  4622.000000  4622.000000         4622.000000     4622.000000   
mean      0.607529     0.224362            0.302250        0.065340   
std       0.488353     0.417206            0.459283        0.247151   
min       0.000000     0.000000            0.000000        0.000000   
25%       0.000000     0.000000            0.000000        0.000000   
50%       1.000000     0.000000            0.000000        0.000000   
75%       1.000000     0.000000            1.000000        0.000000   
max       1.000000     1.000000            1.000000        1.000000   

       Pinto Beans   Sour Cream        Apple         Rice    Coca Cola  \
count  4622.000000  4622.000000  4622.000000  4622.000000  4622.000000   
mean      0.125920     0.370186     0.001298     0.519688     0.005625   
std       0.331795     0.482907     0.036010     0.499666     0.074799   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     1.000000     0.000000   
75%       0.000000     1.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

       Tomatillo Green Chili (Medium)      Lettuce   White Rice     Lemonade  \
count                     4622.000000  4622.000000  4622.000000  4622.000000   
mean                         0.000433     0.376893     0.001082     0.007140   
std                          0.020800     0.484660     0.032876     0.084204   
min                          0.000000     0.000000     0.000000     0.000000   
25%                          0.000000     0.000000     0.000000     0.000000   
50%                          0.000000     0.000000     0.000000     0.000000   
75%                          0.000000     1.000000     0.000000     0.000000   
max                          1.000000     1.000000     1.000000     1.000000   

       Roasted Chili Corn (Medium)  Pineappl

# Inference:
1. From the dataset, we can see that at one time, in a single order, the minimum dishes ordered were 1 and maximum dishes ordered were 15.
2. 25% chance is there that people offered only 1 dish
3. 50% chance is there that people offered <= 8 dishes at one one time
4. 75% chance is there that people offered <= 9 dishes at one time<br>
    CONCLUSION => Most of the time people offered just 1 dish
5. A very obvious observation: As the no. of dishes increased, the item price also increases.
6. We can also see from the 75% row that among different items, the items which were in high demand were:<br>
    a. Rice<br>
    b. Salsa<br>
    c. Sour Cream<br>
    d. Lettuce<br>
    e. Cheese<br>
    f. Black Beans

For the proof, let's calculate their sum column wise:

In [12]:
order_count = data1[final_unique_items].sum(axis=0).sort_values()
order_count

Braised Carnitas                           1
Brown Rice                                 2
Braised Barbacoa                           2
Tomatillo Green Chili (Medium)             2
Vegetarian Black Beans                     3
Tomatillo Red Chili (Hot)                  3
Roasted Chili Corn (Medium)                3
Grapefruit                                 4
Adobo-Marinated and Grilled Steak          4
Adobo-Marinated and Grilled Chicken        5
White Rice                                 5
Pomegranate Cherry                         5
Apple                                      6
Cilantro-Lime Rice                         7
Blackberry                                 8
Clementine                                 8
Peach Orange                               8
Pineapple Orange Banana                    8
Diet Dr. Pepper                           13
Mountain Dew                              15
Nestea                                    20
Coca Cola                                 26
Dr. Pepper

### Now, let's filter the items which have orders >= 500. 
** We kept 500 number of orders for an item as threshold to mark it as a significant order

In [13]:
significant_orders = order_count[order_count > 500]
significant_orders

Pinto Beans                  582
Pinto Beans                  582
Fajita Vegetables            722
Fajita Vegetables            722
Roasted Chili Corn Salsa     727
Guacamole                   1037
Guacamole                   1037
Black Beans                 1345
Black Beans                 1345
Fresh Tomato Salsa          1397
Sour Cream                  1711
Sour Cream                  1711
Lettuce                     1742
Lettuce                     1742
Cheese                      2281
Cheese                      2281
Rice                        2402
Rice                        2402
Salsa                       2808
Salsa                       2808
dtype: int64

## Thus, below items were ordered most:

In [14]:
pd.DataFrame(data=significant_orders, columns=["#Orders"])

#Orders
Pinto Beans                   582
Pinto Beans                   582
Fajita Vegetables             722
Fajita Vegetables             722
Roasted Chili Corn Salsa      727
Guacamole                    1037
Guacamole                    1037
Black Beans                  1345
Black Beans                  1345
Fresh Tomato Salsa           1397
Sour Cream                   1711
Sour Cream                   1711
Lettuce                      1742
Lettuce                      1742
Cheese                       2281
Cheese                       2281
Rice                         2402
Rice                         2402
Salsa                        2808
Salsa                        2808

## Thus, we can conclude that the most favourite dish is "Salsa" with 2808 orders, and least ordered dish is "Braised Carnitas" with just 1 order!

## NOTE: The above result doesn't represent the complete order. They just represents that in maximum orders, these items are used most of the times.

## Now, let's figure out what are the different types of orders and which orders are placed most.

In [15]:
order_item = data1["item_name"].value_counts().sort_values()
order_item

Carnitas Salad                             1
Chips and Mild Fresh Tomato Salsa          1
Veggie Crispy Tacos                        1
Crispy Tacos                               2
Bowl                                       2
Salad                                      2
Steak Salad                                4
Carnitas Salad Bowl                        6
Burrito                                    6
Veggie Salad                               6
Carnitas Crispy Tacos                      7
Veggie Soft Tacos                          7
Chicken Salad                              9
Barbacoa Salad Bowl                       10
Barbacoa Crispy Tacos                     11
Chips and Roasted Chili-Corn Salsa        18
Veggie Salad Bowl                         18
Chips and Tomatillo-Red Chili Salsa       20
Izze                                      20
Chips and Roasted Chili Corn Salsa        22
Barbacoa Soft Tacos                       25
Nantucket Nectar                          27
Steak Sala

## Now, let's set the threshold value for the no. of orders as the mean value of the above table

In [16]:
threshold_order_value = order_item.mean()
threshold_order_value

92.44

## Thus, the number of orders > 92.44 will be considered as the significant orders and should be focussed more.

In [17]:
significant_orders = order_item[order_item > threshold_order_value]
pd.DataFrame(significant_orders)

item_name
Veggie Burrito                       95
Side of Chips                       101
Canned Soda                         104
Chicken Salad Bowl                  110
Chips and Fresh Tomato Salsa        110
Chicken Soft Tacos                  115
Bottled Water                       162
Steak Bowl                          211
Chips                               211
Canned Soft Drink                   301
Steak Burrito                       368
Chips and Guacamole                 479
Chicken Burrito                     553
Chicken Bowl                        726

# So we can see that Chicken Bowl is ordered maximum number of times (726) and Chips and Mild Fresh Tomato Salsa  is ordered just 1 time

# We can also see in how many different ways chicken bowl is ordered!
### For e.g.

In [18]:
ChickenBowl = data1[["item_name", "choice_description", "choice_description_count", "item_price"]][data1["item_name"] == "Chicken Bowl"]
ChickenBowl.sort_values(by="item_price")

item_name                                 choice_description  \
2918  Chicken Bowl  [Fresh Tomato (Mild), [Rice, Sour Cream, Cheese]]   
3664  Chicken Bowl  [Fresh Tomato (Mild), [Lettuce, Fajita Veggies...   
3101  Chicken Bowl  [Fresh Tomato (Mild), [Lettuce, Fajita Veggies...   
2548  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Black...   
3424  Chicken Bowl  [[Roasted Chili Corn Salsa (Medium), Tomatillo...   
2543  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
1609  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Chees...   
1610  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Fajit...   
2511  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Chees...   
1651  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Rice, Che...   
937   Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Fajit...   
934   Chicken Bowl  [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice...   
3497  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
995   Chicken Bowl  [[Fresh Tomato Salsa (Mild), Tomatillo-Red Chi...   
3510  Chicken Bowl  [[Roasted Chili Corn Salsa (Medium), Tomatillo...   
1645  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Tomatillo-Red Chi...   
1649  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
3396  Chicken Bowl  [Tomatillo-Green Chili Salsa (Medium), [Pinto ...   
2062  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...   
1660  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
890   Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Pinto Beans...   
3586  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Black Bea...   
2487  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
3685  Chicken Bowl  [[Roasted Chili Corn Salsa (Medium), Tomatillo...   
2478  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...   
3717  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
3725  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...   
1703  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...   
3811  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...   
3812  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Pinto Beans...   
773   Chicken Bowl  [Tomatillo-Green Chili Salsa (Medium), Black B...   
772   Chicken Bowl  [Tomatillo-Green Chili Salsa (Medium), Black B...   
1705  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Pinto Beans...   
3850  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Pinto Beans...   
1000  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Rice, Che...   
1001  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Black Bea...   
3383  Chicken Bowl  [Fresh Tomato Salsa (Mild), [Rice, Black Beans...   
3359  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...   
3051  Chicken Bowl  [[Tomatillo-Red Chili Salsa (Hot), Tomatillo-G...   
1303  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Tomatillo-Red Chi...   
1307  Chicken Bowl  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...   
3023  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...   
3019  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Cheese, Let...   
2813  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
3053  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Rice, Faj...   
3003  Chicken Bowl  [[Roasted Chili Corn Salsa (Medium), Tomatillo...   
2971  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Black...   
2959  Chicken Bowl  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...   
1405  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Rice, Che...   
2952  Chicken Bowl  [Roasted Chili Corn Salsa (Medium), [Rice, Faj...   
2945  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...   
1432  Chicken Bowl  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...   
2973  Chicken Bowl  [Tomatillo-Red Chili Salsa (Hot), [Rice, Black...   
3076  Chicken Bowl  [Fresh Tomato Salsa (Mild), [Black Beans, Rice.

## Here we can see all the 726 ways in which Chicken Bowl can be ordered, along with the item price for each order type.

## Now, for Chicken Bowl, we can do so many things like:

In [19]:
ChickenBowl.describe()

choice_description_count  item_price
count                726.000000  726.000000
mean                  12.391185   10.113953
std                    3.195882    2.489791
min                    3.000000    8.190000
25%                   11.000000    8.750000
50%                   13.000000    8.750000
75%                   15.000000   11.250000
max                   22.000000   32.940000

## From the above table, we can infer that:
1. min no. of items that can be ordered in a chicken bowl is 3 and max can be 22
2. 75% of the times, the number of items ordered in the Chicken Bowl is <= 15
3. On average, 12 items are ordered in a Chicken Bowl.
4. The average price of the Chicken Bowl is 10.11 dollars
5. 75% of time, the Chicken Bowl price is < 11.25 dollars
6. Very less chances are there when Chick Bowl prices rise to 32.9 dollars. From the table, we see that it is when the choice_description_count is maximum (22)

# Now, within Chicken Bowl, let's see which item is favoured most...

In [20]:
ChickenBowlItems = data1[["item_name",  "item_price"]+final_unique_items][data1["item_name"] == "Chicken Bowl"].sort_values("item_price")
ChickenBowlItems.head()

item_name  item_price  Brown Rice  Black Beans  Fajita Vegetables  \
2918  Chicken Bowl        8.19           0            0                  0   
3664  Chicken Bowl        8.19           0            0                  0   
3101  Chicken Bowl        8.19           0            1                  0   
2548  Chicken Bowl        8.49           0            1                  0   
3424  Chicken Bowl        8.49           0            0                  0   

      Braised Barbacoa  Tomatillo-Red Chili Salsa (Hot)  \
2918                 0                                0   
3664                 0                                0   
3101                 0                                0   
2548                 0                                1   
3424                 0                                0   

      Fresh Tomato Salsa (Mild)  Salsa  Guacamole  Fresh Tomato Salsa  \
2918                          0      0          0                   0   
3664                          0      0          0                   0   
3101                          0      0          0                   0   
2548                          0      1          0                   0   
3424                          0      1          0                   0   

      Fajita Veggies  Salsa  Pinto Beans  Sour Cream  Apple  Rice  Coca Cola  \
2918               0      0            0           1      0     1          0   
3664               1      0            1           1      0     1          0   
3101               1      0            0           1      0     1          0   
2548               0      1            0           0      0     1          0   
3424               1      1            0           1      0     1          0   

      Tomatillo Green Chili (Medium)  Lettuce  White Rice  Lemonade  \
2918                               0        0           0         0   
3664                               0        1           0         0   
3101                               0        1           0         0   
2548                               0        1           0         0   
3424                               0        1           0         0   

      Roasted Chili Corn (Medium)  Tomatillo-Red Chili Salsa (Hot)  \
2918                            0                                0   
3664                            0                                0   
3101                            0                                0   
2548                            0                                1   
3424                            0                                0   

      Pineapple Orange Banana  Guacamole  Roasted Chili Corn Salsa  \
2918                        0          0                         0   
3664                        0          0                         0   
3101                        0          0                         0   
2548                        0          0                         0   
3424                        0          0                         1   

      Roasted Chili Corn Salsa (Medium)  Nestea  Diet Dr. Pepper  Black Beans  \
2918                                  0       0                0            0   
3664                                  0       0                0            0   
3101                                  0       0                0            1   
2548                                  0       0                0            1   
3424                                  1       0                0            0   

      Adobo-Marinated and Grilled Steak  Grapefruit  \
2918                                  0           0   
3664                                  0           0   
3101                                  0           0   
2548                                  0           0   
3424                                  0           0   

      Roasted Chili Corn Salsa (Medium)  Blackberry  Peach Orange  \
2918                                  0           0             0   
3664                                  0 

In [21]:
ChickenBowlItems[final_unique_items].sum(axis=0).sort_values()

Brown Rice                                0
Adobo-Marinated and Grilled Chicken       0
Coca Cola                                 0
Tomatillo Green Chili (Medium)            0
Cilantro-Lime Rice                        0
Lemonade                                  0
Roasted Chili Corn (Medium)               0
Apple                                     0
Pineapple Orange Banana                   0
Diet Coke                                 0
Nestea                                    0
Diet Dr. Pepper                           0
Peach Orange                              0
Adobo-Marinated and Grilled Steak         0
Grapefruit                                0
Dr. Pepper                                0
Pomegranate Cherry                        0
Tomatillo Red Chili (Hot)                 0
Blackberry                                0
Clementine                                0
Braised Barbacoa                          0
Mountain Dew                              0
Coke                            

# Final Inference for Chicken Bowl
=> We see that the **Chicken Bowl** is ordered most and within **Chicken Bowl**, the item which is very much in demand is  **Rice** and **Salsa**

# Beer dataset

In [22]:
raw_data2.head(20)

name  calories  sodium  alcohol  cost
0              Budweiser       144      15      4.7  0.43
1                Schlitz       151      19      4.9  0.43
2              Lowenbrau       157      15      0.9  0.48
3            Kronenbourg       170       7      5.2  0.73
4               Heineken       152      11      5.0  0.77
5          Old_Milwaukee       145      23      4.6  0.28
6             Augsberger       175      24      5.5  0.40
7   Srohs_Bohemian_Style       149      27      4.7  0.42
8            Miller_Lite        99      10      4.3  0.43
9        Budweiser_Light       113       8      3.7  0.40
10                 Coors       140      18      4.6  0.44
11           Coors_Light       102      15      4.1  0.46
12        Michelob_Light       135      11      4.2  0.50
13                 Becks       150      19      4.7  0.76
14                 Kirin       149       6      5.0  0.79
15     Pabst_Extra_Light        68      15      2.3  0.38
16                 Hamms       139      19      4.4  0.43
17   Heilemans_Old_Style       144      24      4.9  0.43
18   Olympia_Goled_Light        72       6      2.9  0.46
19         Schlitz_Light        97       7      4.2  0.47

In [23]:
data2 = raw_data2.copy()

In [24]:
# Print total records and type of variables

data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      20 non-null     object 
 1   calories  20 non-null     int64  
 2   sodium    20 non-null     int64  
 3   alcohol   20 non-null     float64
 4   cost      20 non-null     float64
dtypes: float64(2), int64(2), object(1)
memory usage: 928.0+ bytes


In [25]:
data2.describe(include="all")

name    calories     sodium    alcohol       cost
count            20   20.000000  20.000000  20.000000  20.000000
unique           20         NaN        NaN        NaN        NaN
top     Coors_Light         NaN        NaN        NaN        NaN
freq              1         NaN        NaN        NaN        NaN
mean            NaN  132.550000  14.950000   4.240000   0.494500
std             NaN   30.283702   6.581273   1.087924   0.144858
min             NaN   68.000000   6.000000   0.900000   0.280000
25%             NaN  110.250000   9.500000   4.175000   0.427500
50%             NaN  144.000000  15.000000   4.600000   0.435000
75%             NaN  150.250000  19.000000   4.900000   0.485000
max             NaN  175.000000  27.000000   5.500000   0.790000

## Inferences from the above table
1. We see that in the data set, there are total 20 records, and each record represents a special type of bear, since unique row of name is equal to count row
2. From mean row, we see that on average, the bear has 132.55 Calories, 14.95 Sodium and 4.24 Alcohol, and on average, a bear costs 0.4945
3. NOTE: For Alcohol, the std dev is minimum: 0.144858, while for other items like Calories and Sodium, the standard deviation is very high ==> The Alcohol is more or less in same amount in any type of bear, but the amount of Calories and amount of Sodium can vary significantly from beer to beer. 
3. From 50% row, we see 144 for Calories, 15 for Sodium and 0.45 for Alcohol
    * <b>Inference#1</b> mean for Calories is (144 - 132.55 = 11.45) < less than the 50% percentile bear, but also < SD for calories(30.283702), thus we can say that slightly more bears have Calories < mean(132.55)
    * <b>Inference#2</b> mean for Sodium is (15 - 14.95 = 0.05) > than the 50% percentile bear and also very much less than the standard deviation(6.581273), thus we can say that the no. of bears with sodium < 14.95 is same as the no. of bears with sodium > 14.95
    * <b>Inference#3</b> mean for Alcohol is (4.6 - 4.24 = 0.36) < than the 50% percentile bear and also very much less than the standard deviation(1.08), thus we can say that the no. of bears with sodium < 4.24 is approximately same as the no. of bears with sodium > 14.95
    * <b>Inference#4</b> mean for Cost is (0.494500 - 0.435 = 0.06) > than the 50% percentile bear and also very much less than the standard deviation(0.144), thus we can say that the no. of bears with sodium < 0.4945 is approximately same as the no. of bears with sodium > 14.95

In [26]:
data2.shape

(20, 5)

In [27]:
# How is the dataset indexed?
data2.index

RangeIndex(start=0, stop=20, step=1)

In [28]:
# Which item has most calories and what what does it cost?

cal = data2.groupby("name").sum()
cal = cal.sort_values("calories", ascending = False)
cal.head(1)

calories  sodium  alcohol  cost
name                                       
Augsberger       175      24      5.5   0.4

In [29]:
# Which bear has highest alcohol?
alcohol = data2.groupby("name").sum().sort_values("alcohol", ascending=False)
alcohol.head(1)

calories  sodium  alcohol  cost
name                                       
Augsberger       175      24      5.5   0.4

In [30]:
# Which is the most expensive beer?

expensive = data2.sort_values("cost", ascending=False)
expensive.head(1)

name  calories  sodium  alcohol  cost
14  Kirin       149       6      5.0  0.79

In [31]:
# What is the cost per calorie for each beer? Which beer has
# highest cost per calorie?

cost_per_cal = data2["cost"]/data2["calories"]
data2["Cost_per_cal"] = cost_per_cal

cpc = data2.sort_values("Cost_per_cal", ascending=False)
cpc.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal
18  Olympia_Goled_Light        72       6      2.9  0.46      0.006389

In [32]:
# Which beer has highest cost per sodium?

cost_per_sod = data2["cost"]/data2["sodium"]
data2["cost_per_sod"] = cost_per_sod
cps = data2.sort_values("cost_per_sod", ascending=False)
cps.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal  cost_per_sod
14  Kirin       149       6      5.0  0.79      0.005302      0.131667

In [33]:
# Which beer has highest cost per alcohol?
cost_per_alc = data2["cost"]/data2["alcohol"]
data2["cost_per_alc"] = cost_per_alc
cpa = data2.sort_values("cost_per_alc", ascending=False)
cpa.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal  cost_per_sod  \
2  Lowenbrau       157      15      0.9  0.48      0.003057         0.032   

   cost_per_alc  
2      0.533333

In [34]:
# Thus highest cost per calories is:

cpc.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal
18  Olympia_Goled_Light        72       6      2.9  0.46      0.006389

In [35]:
# Thus highest cost per sodium is:

cps.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal  cost_per_sod
14  Kirin       149       6      5.0  0.79      0.005302      0.131667

In [36]:
# Thus highest cost per alcohol is:

cpa.head(1)

name  calories  sodium  alcohol  cost  Cost_per_cal  cost_per_sod  \
2  Lowenbrau       157      15      0.9  0.48      0.003057         0.032   

   cost_per_alc  
2      0.533333

In [37]:
names == name if name in ("Olympia_Goled_Light", "Kirin", "Lowenbrau") else False

NameError: name 'name' is not defined

In [ ]:
# Let's try to filter above data using pandas
names = data2["name"]

validity = [True if val in ("Olympia_Goled_Light", "Kirin", "Lowenbrau") else False for val in (names)]

extremum = data2[validity]
extremum

In [ ]:
# Now, let's sort the extremum from the most to less expensive
extremum.sort_values("cost", ascending=False, inplace=True)
extremum

In [ ]:
data3 = raw_data3.copy()

In [ ]:
data3.head()

In [ ]:
data3.tail()

In [ ]:
data3.info()

In [ ]:
stats = data3.describe().sort_values("mean",axis=1)
stats

In [ ]:
# Exclude some columns from the result:
final_stats = stats[stats.columns.difference(['Year', 'Total', 'Population'])].sort_values(by="mean", axis=1)

final_stats

# Inferences:
1. From mean row, we can see that most of the crime takes place in Property and Murder has got the least count.

In [ ]:
murder_cases = data3.groupby("Murder").mean().sort_values("Murder")[["Year", "Population"]]
rape_cases = data3.groupby("Forcible_Rape").mean().sort_values("Forcible_Rape")[["Year", "Population"]]
robbery_cases = data3.groupby("Robbery").mean().sort_values("Robbery")[["Year", "Population"]]
Aggravated_assault_cases = data3.groupby("Aggravated_assault").mean().sort_values("Aggravated_assault")[["Year", "Population"]]
Vehicle_Theft = data3.groupby("Vehicle_Theft").mean().sort_values("Vehicle_Theft")[["Year", "Population"]]
Violent = data3.groupby("Violent").mean().sort_values("Violent")[["Year", "Population"]]
Burglary = data3.groupby("Burglary").mean().sort_values("Burglary")[["Year", "Population"]]
Larceny_Theft = data3.groupby("Larceny_Theft").mean().sort_values("Larceny_Theft")[["Year", "Population"]]
Property = data3.groupby("Property").mean().sort_values("Property")[["Year", "Population"]]

In [ ]:
murder_cases

## Inference on Murder:
1. From the above table, we can infer that as the population increases, the murder cases sometimes increase and sometimes decrease, thus saying that murder cases increases with increasing population would be False.

In [ ]:
rape_cases

In [ ]:
data3[["Forcible_Rape", "Year", "Population"]].corr()

## Inference for Forcible_Rape

From the above 2 tables, we can infer that:
1. As population increases, Forcible_Rape case also generally increases, and because population increases by Year, thus it would not be wrong to say that Rape cases have increased as time has passed.

In [ ]:
robbery_cases

In [ ]:
data3[["Robbery", "Year", "Population"]].corr()

## Inference for Robbery cases:
1. We can see that population and time has no significant impact on Robbery cases

In [ ]:
Aggravated_assault_cases

In [ ]:
data3[["Aggravated_assault", "Year", "Population"]].corr()

## Inference for Aggravated assault
1. As population has increased, these cases have increased in number.

In [ ]:
Vehicle_Theft

In [ ]:
data3[["Vehicle_Theft", "Year", "Population"]].corr()

## Inference for Vehicle Theft
1. It has nothing to do with the population

In [ ]:
Violent

In [ ]:
data3[["Violent", "Year", "Population"]].corr()

## Inference for Violence:
1. As population has increased, Violence has also increased.

In [ ]:
Burglary

In [ ]:
data3[["Burglary", "Year", "Population"]].corr()

## Inference for Burglary:
1. Burglary has nothing to do with the population rise.

In [ ]:
Larceny_Theft

In [ ]:
data3[["Larceny_Theft", "Year", "Population"]].corr()

## Inference for Larency_Theft
1. Whth increase in population, Larency Theft has generally increased.

In [ ]:
Property

In [ ]:
data3[["Property", "Year", "Population"]].corr()

## Inference for Property Crime
1. Doesn't depend much on population

In [ ]:
data4 = raw_data4.copy()

In [ ]:
data4.shape

In [ ]:
data4.head()

In [ ]:
for item, csv, doc in data4[["Item", "CSV", "Doc"]].values:
    print(item, csv, doc)

## No furthur breakdown makes sense to me, as this table just represents the information about the csv files.